### PR TITLE
New output plugin "flowcount"

### PR DESCRIPTION
**Pull request by nokute78**
I added a new output plugin "out_fcount".
It is inspired by
https://github.com/tagomoris/fluent-plugin-flowcounter
https://github.com/sonots/fluent-plugin-flowcounter-simple

## Usecase

To count records and data size per second/minute(default)/hour/day.
Simple Benchmark.

## How to use

You can set property `unit`.

`$ bin/fluent-bit -i cpu -o fcount -p unit=second`

## Example

```python
$ bin/fluent-bit -i cpu -t hoge -o fcount -m hoge -o stdout -m hoge
Fluent-Bit v0.10.0
Copyright (C) Treasure Data

[2016/11/09 23:39:02] [ info] [engine] started
[0] hoge: [1478702343, {"cpu_p"=>2.000000, "user_p"=>1.000000, "system_p"=>1.000000, "cpu0.p_cpu"=>2.000000, "cpu0.p_user"=>1.000000, "cpu0.p_system"=>1.000000}]
[1] hoge: [1478702344, {"cpu_p"=>6.000000, "user_p"=>5.000000, "system_p"=>1.000000, "cpu0.p_cpu"=>6.000000, "cpu0.p_user"=>5.000000, "cpu0.p_system"=>1.000000}]
[2] hoge: [1478702345, {"cpu_p"=>2.000000, "user_p"=>2.000000, "system_p"=>0.000000, "cpu0.p_cpu"=>2.000000, "cpu0.p_user"=>2.000000, "cpu0.p_system"=>0.000000}]
[3] hoge: [1478702346, {"cpu_p"=>5.000000, "user_p"=>4.000000, "system_p"=>1.000000, "cpu0.p_cpu"=>5.000000, "cpu0.p_user"=>4.000000, "cpu0.p_system"=>1.000000}]
[0] hoge: [1478702347, {"cpu_p"=>3.000000, "user_p"=>3.000000, "system_p"=>0.000000, "cpu0.p_cpu"=>3.000000, "cpu0.p_user"=>3.000000, "cpu0.p_system"=>0.000000}]
.
.
[0] hoge: [1478702397, {"cpu_p"=>0.000000, "user_p"=>0.000000, "system_p"=>0.000000, "cpu0.p_cpu"=>0.000000, "cpu0.p_user"=>0.000000, "cpu0.p_system"=>0.000000}]
[1] hoge: [1478702398, {"cpu_p"=>1.000000, "user_p"=>1.000000, "system_p"=>0.000000, "cpu0.p_cpu"=>1.000000, "cpu0.p_user"=>1.000000, "cpu0.p_system"=>0.000000}]
[2] hoge: [1478702399, {"cpu_p"=>0.000000, "user_p"=>0.000000, "system_p"=>0.000000, "cpu0.p_cpu"=>0.000000, "cpu0.p_user"=>0.000000, "cpu0.p_system"=>0.000000}]
[3] hoge: [1478702400, {"cpu_p"=>1.000000, "user_p"=>0.000000, "system_p"=>1.000000, "cpu0.p_cpu"=>1.000000, "cpu0.p_user"=>0.000000, "cpu0.p_system"=>1.000000}]
[4] hoge: [1478702401, {"cpu_p"=>1.000000, "user_p"=>1.000000, "system_p"=>0.000000, "cpu0.p_cpu"=>1.000000, "cpu0.p_user"=>1.000000, "cpu0.p_system"=>0.000000}]
[out_fcount] hoge:{"count":60, "bytes":7560, "count/minute":1, "bytes/minute":126}
```

### CODE REVIEW

File: plugins/out_fcount/out_fcount.c [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+    unit = flb_output_get_property("unit", ins);
+    if ( unit == NULL ) {
+        /* using default */
+        return 0;
+    }
+
+    flb_debug("[%s]unit is \"%s\"",PLUGIN_NAME, unit);
+
+    if ( !strcasecmp(unit, FLB_UNIT_SEC) ) {
```
Comment (edsiper) [inline plugins/out_fcount/out_fcount.c]: same as above (spaces)

File: plugins/out_fcount/out_fcount.c [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    ctx->counts++;
+    ctx->bytes += size;
+    /*TODO parse obj and count up specific data */
+}
+
+static time_t get_timestamp_from_msgpack(msgpack_object *p)
+{
+    if ( p != NULL && p->via.array.size != 0 ) {
+        return p->via.array.ptr[0].via.u64; /* FIXME */
+    }else{
```
Comment (edsiper) [inline plugins/out_fcount/out_fcount.c]: please do the _else_ always in a new line, e.g:

```
if () {

}
else {

}
```

File: plugins/out_fcount/out_fcount.c [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    count_initialized(ctx);
+    ctx->last_checked = time(NULL);
+
+    flb_output_set_context(ins, ctx);
+
+    return 0;
+}
+
+static void out_fcount_flush(void *data, size_t bytes,
+                     char *tag, int tag_len,
```
Comment (edsiper) [inline plugins/out_fcount/out_fcount.c]: align parameters to the top line, e.g:

```
static void out_fcount_flush(void *data, size_t bytes,
                                             char *tag, int tag_len,
```

**Comment by edsiper** [comment]
thanks, I added some comments inside the code, once that minor code-style fixes are done we can merge.

**Comment by repeatedly** [comment]
fcount is unclear for me. Why don't you use flowcount?

**Comment by kzk** [comment]
I would much prefer using 'consistent' name across fluentd / fluent-bit. So using same name 'flowcounter' is more natural for fluentd users I believe.

**Comment by nokute78** [comment]
@edsiper 
Thank you for your reviewing. I will fix.

@repeatedly  @kzk 
> Why don't you use flowcount?
I think so.  I will also fix it.

The reason is that I follow the naming convention of Elasticsearch plugin.
We can use that plugin with short option "es". (e.g. $ bin/fluent-bit -i cpu -o es)
So, I used short name option "fcount". (and name is "flowcount")

**Comment by nokute78** [comment]
I committed.

**Comment by edsiper** [comment]
thanks for the changes, merged!

**Pull request merged**